File: archive/pulumi-ts/Pr5410/lib/database-stack.ts
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export interface DatabaseStackArgs {
  environmentSuffix: string;
  subnetIds: pulumi.Output<string[]>;
  securityGroupId: pulumi.Output<string>;
  dbSecretArn?: string;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class DatabaseStack extends pulumi.ComponentResource {
  public readonly dbEndpoint: pulumi.Output<string>;
  public readonly dbInstanceId: pulumi.Output<string>;

  constructor(
    name: string,
    args: DatabaseStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:database:DatabaseStack', name, args, opts);

    // Create KMS key for RDS encryption
    const kmsKey = new aws.kms.Key(
      `payment-db-kms-${args.environmentSuffix}`,
      {
        description: 'KMS key for RDS encryption',
        enableKeyRotation: true,
        tags: pulumi.all([args.tags]).apply(([t]) => ({
          ...t,
          Name: `payment-db-kms-${args.environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    new aws.kms.Alias(
      `payment-db-kms-alias-${args.environmentSuffix}`,
      {
        name: `alias/payment-db-${args.environmentSuffix}`,
        targetKeyId: kmsKey.keyId,
      },
      { parent: this }
    );

    // Create DB subnet group
    const dbSubnetGroup = new aws.rds.SubnetGroup(
      `payment-db-subnet-group-${args.environmentSuffix}`,
      {
        subnetIds: args.subnetIds,
        tags: pulumi.all([args.tags]).apply(([t]) => ({
          ...t,
          Name: `payment-db-subnet-group-${args.environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    // Fetch database credentials from Secrets Manager
    let dbUsername: pulumi.Input<string> = 'postgres';
    let dbPassword: pulumi.Output<string> = pulumi.secret(
      'temporarypassword123'
    );

    if (args.dbSecretArn) {
      const dbSecret = pulumi.output(
        aws.secretsmanager.getSecretVersion({
          secretId: args.dbSecretArn,
        })
      );

      const secretString = dbSecret.apply(s => JSON.parse(s.secretString));
      dbUsername = secretString.apply(
        (s: { username?: string; password: string }) => s.username || 'postgres'
      );
      dbPassword = pulumi.secret(
        secretString.apply(
          (s: { username?: string; password: string }) => s.password
        )
      ) as pulumi.Output<string>;
    }

    // Create RDS PostgreSQL instance
    const dbInstance = new aws.rds.Instance(
      `payment-db-${args.environmentSuffix}`,
      {
        identifier: `payment-db-${args.environmentSuffix}`,
        engine: 'postgres',
        engineVersion: '15.14',
        instanceClass: 'db.t3.micro',
        allocatedStorage: 20,
        storageType: 'gp3',
        storageEncrypted: true,
        kmsKeyId: kmsKey.arn,

        dbName: 'paymentdb',
        username: dbUsername,
        password: dbPassword,

        dbSubnetGroupName: dbSubnetGroup.name,
        vpcSecurityGroupIds: [args.securityGroupId],

        multiAz: true,
        publiclyAccessible: false,

        backupRetentionPeriod: 7,
        backupWindow: '03:00-04:00',
        maintenanceWindow: 'mon:04:00-mon:05:00',

        skipFinalSnapshot: true,
        deletionProtection: false, // Must be destroyable for testing

        enabledCloudwatchLogsExports: ['postgresql', 'upgrade'],

        tags: pulumi.all([args.tags]).apply(([t]) => ({
          ...t,
          Name: `payment-db-${args.environmentSuffix}`,
        })),
      },
      { parent: this }
    );

    this.dbEndpoint = dbInstance.endpoint;
    this.dbInstanceId = dbInstance.id;

    this.registerOutputs({
      dbEndpoint: this.dbEndpoint,
      dbInstanceId: this.dbInstanceId,
    });
  }
}

File: archive/cdktf-py/Pr6435/lib/database_stack.py
"""Database Stack - RDS PostgreSQL with Multi-AZ deployment."""

from constructs import Construct
from cdktf_cdktf_provider_aws.db_subnet_group import DbSubnetGroup
from cdktf_cdktf_provider_aws.db_instance import DbInstance
from cdktf_cdktf_provider_aws.ssm_parameter import SsmParameter
from cdktf_cdktf_provider_aws.kms_key import KmsKey
from cdktf_cdktf_provider_aws.kms_alias import KmsAlias
import json


class DatabaseStack(Construct):
    """Database infrastructure for payment processing application."""

    def __init__(
        self,
        scope: Construct,
        construct_id: str,
        *,
        environment_suffix: str,
        vpc_id: str,
        private_subnet_ids: list,
        database_security_group_id: str,
        **kwargs
    ):
        """Initialize database stack."""
        super().__init__(scope, construct_id)

        # KMS Key for RDS Encryption
        kms_key = KmsKey(
            self,
            "rds_kms_key",
            description=f"KMS key for RDS encryption - {environment_suffix}",
            deletion_window_in_days=10,
            enable_key_rotation=True,
            policy=json.dumps({
                "Version": "2012-10-17",
                "Statement": [
                    {
                        "Sid": "Enable IAM User Permissions",
                        "Effect": "Allow",
                        "Principal": {
                            "AWS": "*"
                        },
                        "Action": "kms:*",
                        "Resource": "*"
                    }
                ]
            }),
            tags={
                "Name": f"payment-rds-kms-{environment_suffix}",
            },
        )

        KmsAlias(
            self,
            "rds_kms_alias",
            name=f"alias/payment-rds-{environment_suffix}",
            target_key_id=kms_key.key_id,
        )

        # DB Subnet Group
        db_subnet_group = DbSubnetGroup(
            self,
            "db_subnet_group",
            name=f"payment-db-subnet-group-{environment_suffix}",
            description=f"Subnet group for payment database - {environment_suffix}",
            subnet_ids=private_subnet_ids,
            tags={
                "Name": f"payment-db-subnet-group-{environment_suffix}",
            },
        )

        # RDS PostgreSQL Instance
        self._rds_instance = DbInstance(
            self,
            "rds_instance",
            identifier=f"payment-db-{environment_suffix}",
            engine="postgres",
            engine_version="15",
            instance_class="db.t3.micro",
            allocated_storage=20,
            storage_type="gp3",
            storage_encrypted=True,
            kms_key_id=kms_key.arn,
            db_name="paymentdb",
            username="dbadmin",
            manage_master_user_password=True,
            multi_az=True,
            db_subnet_group_name=db_subnet_group.name,
            vpc_security_group_ids=[database_security_group_id],
            backup_retention_period=7,
            backup_window="03:00-04:00",
            maintenance_window="mon:04:00-mon:05:00",
            skip_final_snapshot=True,
            deletion_protection=False,
            enabled_cloudwatch_logs_exports=["postgresql", "upgrade"],
            performance_insights_enabled=True,
            performance_insights_retention_period=7,
            auto_minor_version_upgrade=True,
            publicly_accessible=False,
            tags={
                "Name": f"payment-db-{environment_suffix}",
            },
        )

        # Store DB connection string in SSM Parameter Store
        # Note: In production, retrieve password from Secrets Manager
        connection_string = f"postgresql://dbadmin@{self._rds_instance.endpoint}/paymentdb"

        self._db_parameter = SsmParameter(
            self,
            "db_connection_parameter",
            name=f"/payment/{environment_suffix}/db/connection",
            description=f"Database connection string for payment app - {environment_suffix}",
            type="SecureString",
            value=connection_string,
            tags={
                "Name": f"payment-db-connection-{environment_suffix}",
            },
        )

    @property
    def rds_endpoint(self) -> str:
        """Return RDS endpoint."""
        return self._rds_instance.endpoint

    @property
    def rds_arn(self) -> str:
        """Return RDS ARN."""
        return self._rds_instance.arn

    @property
    def db_connection_parameter_name(self) -> str:
        """Return SSM parameter name for DB connection string."""
        return self._db_parameter.name
